extensions/ql-vscode/src/common/vscode/file-path-discovery.ts

```diff
@@ -0,0 +1,203 @@
+import { Discovery } from "../discovery";
+import {
+  EventEmitter,
+  RelativePattern,
+  Uri,
+  WorkspaceFoldersChangeEvent,
+  workspace,
+} from "vscode";
+import { MultiFileSystemWatcher } from "./multi-file-system-watcher";
+import { AppEventEmitter } from "../events";
+import { extLogger } from "..";
+import { lstat, pathExists } from "fs-extra";
+import { containsPath } from "../../pure/files";
+import { getOnDiskWorkspaceFoldersObjects } from "./workspace-folders";
+
+interface PathData {
+  path: string;
+}
+
+/**
+ * Discovers all files matching a given filter contained in the workspace.
+ *
+ * Scans the whole workspace on startup, and then watches for changes to files
+ * to do the minimum work to keep up with changes.
+ *
+ * Can configure which changes it watches for, which files are considered
+ * relevant, and what extra data to compute for each file.
+ */
+export abstract class FilePathDiscovery<T extends PathData> extends Discovery {
```

Comment:
For the record, I did just try removing the inheritance of `Discovery`, but I don't think it was necessarily better.

The idea is that instead of keeping `changedFilePaths` and the `refresh` method, we just process all events immediately. I think this would work, in that we'd be able to process events and keep up with on-disk state just as well.

The problem was it's harder to test because you can't be sure when it's done processing that event. You'd have to introduce a bunch of arbitrary sleeps, which would be a nightmare for speed and flakiness. One benefit that `Discovery` and the `refresh` method gives is that you can easily wait until a refresh is complete.

It also didn't make much difference to line count. The diff was +40 lines and also -40 lines, so completely net-zero.